src/pc/render_shared.c

```diff
@@ -0,0 +1,98 @@
+#include <common.h>
+#include <log.h>
+#include <game.h>
+#include <stdlib.h>
+#include "pc.h"
+#include <assert.h>
+#include "sdl_defs.h"
+
+enum Renderers render_mode = RENDER_SOFT;
+
+int SoftDrawSync(int mode);
+DISPENV* SoftPutDispEnv(DISPENV* env);
+void SoftSetDrawEnv(DR_ENV* dr_env, DRAWENV* env);
+DRAWENV* SoftPutDrawEnv(DRAWENV* env);
+void SoftDrawOTag(OT_TYPE* p);
+
+int GlDrawSync(int mode);
+DISPENV* GlPutDispEnv(DISPENV* env);
+void GlSetDrawEnv(DR_ENV* dr_env, DRAWENV* env);
+DRAWENV* GlPutDrawEnv(DRAWENV* env);
+void GlDrawOTag(OT_TYPE* p);
+
+int MyDrawSync(int mode) {
+    if (render_mode == RENDER_SOFT) {
+        return SoftDrawSync(mode);
+    } else {
+        return GlDrawSync(mode);
+    }
+    return 0;
+}
+
+DISPENV* MyPutDispEnv(DISPENV* env) {
+    if (render_mode == RENDER_SOFT) {
+        return SoftPutDispEnv(env);
+    } else {
+        return GlPutDispEnv(env);
+    }
+    return env;
+}
+
+void MySetDrawEnv(DR_ENV* dr_env, DRAWENV* env) {
+    if (render_mode == RENDER_SOFT) {
+        SoftSetDrawEnv(dr_env, env);
+    } else {
+        GlSetDrawEnv(dr_env, env);
+    }
+}
+
+DRAWENV* MyPutDrawEnv(DRAWENV* env) {
+    if (render_mode == RENDER_SOFT) {
+        return SoftPutDrawEnv(env);
+    } else {
+        return GlPutDrawEnv(env);
+    }
+
+    return env;
+}
+
+void MyDrawOTag(OT_TYPE* p) {
+    if (render_mode == RENDER_SOFT) {
+        SoftDrawOTag(p);
+    } else {
+        GlDrawOTag(p);
+    }
+}
+
+extern u8* GetFb();
+extern u16 g_RawVram[VRAM_W * VRAM_H];
+
+int MyLoadImage(RECT* rect, u_long* p) {
+    DEBUGF("(%X, %X, %X, %X): %p", rect->x, rect->y, rect->w, rect->h, p);
+
+    {
+        u16* mem = (u16*)p;
+        u16* vram = g_RawVram;
+        vram += rect->x + rect->y * VRAM_W;
+
+        for (int i = 0; i < rect->h; i++) {
+            memcpy(vram, mem, rect->w * 2);
+            mem += rect->w;
+            vram += VRAM_W;
+        }
+    }
+
+    // copy to software vram
```

Comment:
Can we do a `if (render_mode == RENDER_SOFT) {` here too? Or do you plan to switch between hw and sw rendering on the fly?

Reply:
Right now it's able to switch with F8, I don't know how useful it is or not. I prefer software rendering for this case so I don't need to be able to switch. What do you prefer?

Reply:
I have yet to try this PR (hence why it's taking so long for me to review). I missed the F8 thingie. That's very cool actually!